src/main.rs

```diff
@@ -62,15 +75,87 @@ async fn main() -> Result<()> {
 		.collect::<Vec<_>>();
 	info!("got environmental variables: {:?}", env_vars);
 
+	let (panel_notifications_fd, daemon_notifications_fd) =
+		notifications::create_socket().expect("Failed to create notification socket");
+	let mut panel_env_vars = env_vars.clone();
+
+	panel_env_vars.push((
+		PANEL_NOTIFICATIONS_FD.to_string(),
+		panel_notifications_fd.as_raw_fd().to_string(),
+	));
+
+	let panel_notifications_fd_pre = Arc::new(Mutex::new(panel_notifications_fd));
+	let span = info_span!(parent: None, "cosmic-panel");
+	let stdout_span = span.clone();
+	let stderr_span = span;
 	process_manager
 		.start(
 			Process::new()
 				.with_executable("cosmic-panel")
-				.with_env(env_vars.clone()),
+				// XXX this should be safe because cosmic-session runs on a single thread
+				.with_fds(move || {
+					let panel_notifications_fd = panel_notifications_fd_pre.clone();
+					let fd = panel_notifications_fd.lock().unwrap();
+					vec![fd.as_raw_fd()]
+				})
+				.with_on_stdout(move |_, _, line| {
+					let stdout_span = stdout_span.clone();
+					async move {
+						info!("{}", line);
+					}
+					.instrument(stdout_span)
+				})
+				.with_on_stderr(move |_, _, line| {
+					let stderr_span = stderr_span.clone();
+					async move {
+						warn!("{}", line);
+					}
+					.instrument(stderr_span)
+				})
+				.with_env(panel_env_vars.clone()),
 		)
 		.await
 		.expect("failed to start panel");
 
+	let mut daemon_env_vars = env_vars.clone();
+	daemon_env_vars.push((
+		DAEMON_NOTIFICATIONS_FD.to_string(),
+		daemon_notifications_fd.as_raw_fd().to_string(),
+	));
+	let span = info_span!(parent: None, "cosmic-notifications");
+	let stdout_span = span.clone();
+	let stderr_span = span;
+	let daemon_notifications_fd_pre = Arc::new(Mutex::new(daemon_notifications_fd));
+
+	process_manager
+		.start(
+			Process::new()
+				.with_executable("cosmic-notifications")
+				// XXX this should be safe because cosmic-session runs on a single thread
+				.with_fds(move || {
+					let daemon_notifications_fd = daemon_notifications_fd_pre.clone();
```

Comment:
Looks like the use of `Arc` and `Mutex` is unnecessary here? It seems to compiling if the definition of `daemon_notifications_fd_pre` above is removed, and this closure is changed to just `vec![daemon_notifications_fd.as_raw_fd()]`. The fd should live as long as the closure exists, so I believe that won't have any issues.